zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentReconstructProcessor.java

```diff
@@ -84,42 +90,71 @@ public void processRecord(final TypedRecord<DeploymentRecord> record) {
           "Deployments are already stored and don't need to be reconstructed");
       return;
     }
+    progressState = ProgressState.fromDeploymentRecord(record.getValue());
 
-    final var key = keyGenerator.nextKey();
-    final var resource = findNextResource();
-    if (resource == null) {
-      stateWriter.appendFollowUpEvent(key, DeploymentIntent.RECONSTRUCTED_ALL, record.getValue());
-      return;
-    }
+    try {
+      final var key = keyGenerator.nextKey();
 
-    final DeploymentRecord deploymentRecord;
-    if (resource.deploymentKey() != NO_DEPLOYMENT_KEY) {
-      final var allResourcesOfDeployment =
-          findResourcesWithDeploymentKey(resource.tenantId(), resource.deploymentKey());
-      deploymentRecord =
-          recreateDeploymentForResources(
-              resource.deploymentKey(), resource.tenantId(), allResourcesOfDeployment);
-    } else {
-      deploymentRecord = createNewDeploymentForResource(resource);
+      final var resource = findNextResource();
+      if (resource == null) {
+        stateWriter.appendFollowUpEvent(key, DeploymentIntent.RECONSTRUCTED_ALL, record.getValue());
+        return;
+      }
+
+      final DeploymentRecord deploymentRecord;
+      if (resource.deploymentKey() != NO_DEPLOYMENT_KEY) {
+        final var allResourcesOfDeployment =
+            findResourcesWithDeploymentKey(resource.tenantId(), resource.deploymentKey());
+        deploymentRecord =
+            recreateDeploymentForResources(
+                resource.deploymentKey(), resource.tenantId(), allResourcesOfDeployment);
+      } else {
+        deploymentRecord = createNewDeploymentForResource(resource);
+      }
+      stateWriter.appendFollowUpEvent(
+          deploymentRecord.getDeploymentKey(), DeploymentIntent.RECONSTRUCTED, deploymentRecord);
+      // trigger reconstruction of another deployment reconstruction
+      cachedDeploymentRecordCommand.reset();
+      cachedDeploymentRecordCommand
+          .setTenantId(deploymentRecord.getTenantId())
+          .setReconstructionKey(deploymentRecord.getDeploymentKey())
+          .setReconstructionProgress(progressState.progress());
+      commandWriter.appendNewCommand(DeploymentIntent.RECONSTRUCT, cachedDeploymentRecordCommand);
+
+    } finally {
+      // progressState contains a reference to a resource, so no need to keep it
+      progressState = null;
     }
-    stateWriter.appendFollowUpEvent(
-        deploymentRecord.getDeploymentKey(), DeploymentIntent.RECONSTRUCTED, deploymentRecord);
-    // trigger reconstruction of another deployment reconstruction
-    commandWriter.appendNewCommand(DeploymentIntent.RECONSTRUCT, cachedDeploymentRecordCommand);
+  }
+
+  @VisibleForTesting
+  ProgressState getProgressState() {
+    return progressState;
```

Comment:
:wrench: This appears to be unused, even in tests. We should remove the method and the `progressState` field in general because it's generally not okay to keep such state in a processor. That also allows us to remove the try-finally block in above.

Reply:
Yes that is a "typo", was used before in test to check that the right progress was saved, but decided to not save the progress state in the state, it's just a temporary variable as it simplified "book-keeping". I'll try to remove it as field.